### PR TITLE
Added a new data source for fetching the details of available alloydb locations

### PR DESCRIPTION
**Pull request by ravisiddhu**
This PR is for https://github.com/hashicorp/terraform-provider-google/issues/13812

A new datasource `google_alloydb_locations` for fetching the details of available alloydb locations. It receives `project`  as arguments and returns the details of all the available alloydb locations.

I have added documentation and tested this new data source with unit tests.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

```release-note:new-datasource
A new datasource for fetching the details of available alloydb locations.
```

### CODE REVIEW

**Comment by ScottSuarez** [comment]
Could you pull the latest version of the provider?

@melinath do you know why validator is failing here? This datasource doesn't seem to be doing anything different than other datasources

**Comment by melinath** [comment]
@ScottSuarez the TFV test step fails if TPG fails to compile (which is the case here). We should probably make the TFV tests run _after_ the TPG unit tests in all cases; this is frequently confusing to folks.

**Comment by ravisiddhu** [comment]
@ScottSuarez  I have pulled the latest version of the provider.

File: mmv1/third_party/terraform/utils/config.go.erb [hunk 1]
```diff
@@ -575,6 +576,16 @@ func (c *Config) NewSqlAdminClient(userAgent string) *sqladmin.Service {
 	return clientSqlAdmin
 }
 
+func (c *Config) NewAlloydbClient() *alloydb.AlloyDBAdminClient {
+	log.Printf("[INFO] Instantiating Alloydb client")
+	clientAlloydb, err := alloydb.NewAlloyDBAdminRESTClient(c.context, option.WithHTTPClient(c.Client))
```
Comment (isaurabhuttam) [inline mmv1/third_party/terraform/utils/config.go.erb:581]: Is it possible to support end-point override functionality in alloydb client also, like other clients?

**Comment by ScottSuarez** [comment]
looks like you may need to upgrade the google api package
```
google-beta/config.go:21:2: no required module provides package cloud.google.com/go/alloydb/apiv1; to add it:
```

**Comment by ravisiddhu** [comment]
I have added the dependency now.

File: mmv1/third_party/terraform/utils/config.go.erb (hunk 1 above)
Comment (ravisiddhu) [reply mmv1/third_party/terraform/utils/config.go.erb:716]: yes

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/go.mod.erb [hunk 2]
```diff
@@ -4,6 +4,7 @@ module github.com/hashicorp/terraform-provider-google<%= "-" + version unless ve
 go 1.18
 
 require (
+	cloud.google.com/go/alloydb v0.2.0
```
Comment (rileykarson) [inline mmv1/third_party/terraform/go.mod.erb:7]: Can we avoid adding new GRPC clients? Let's use https://github.com/googleapis/google-api-go-client or make HTTP calls directly similar to MMv1 instead.

Comment (rileykarson) [reply mmv1/third_party/terraform/go.mod.erb:7]: Sent https://github.com/GoogleCloudPlatform/magic-modules/pull/7544 to update guidance! Sorry, we hadn't considered folks would use a veneer client.

**Review by rileykarson — CHANGES_REQUESTED** [review]
Upgrading request to not add new clients to blocking

**Comment by ravisiddhu** [comment]
> Upgrading request to not add new clients to blocking

sorry, I didn't understand your comment, is there any action item on me?

**Comment by rileykarson** [comment]
See https://github.com/GoogleCloudPlatform/magic-modules/pull/7388#discussion_r1151005465

**Comment by ScottSuarez** [comment]
Hi @ravisiddhu, let me know if you need any help here. Feel free to message me. In alignment with what Riley is saying we'll need to switch to using the go REST client as opposed to gRPC client. The reasoning for this is because our client retry logic doesn't get captured in the gRPC client causing a divergence from current resource behavior.

**Comment by ravisiddhu** [comment]
> Hi @ravisiddhu, let me know if you need any help here. Feel free to message me. In alignment with what Riley is saying we'll need to switch to using the go REST client as opposed to gRPC client. The reasoning for this is because our client retry logic doesn't get captured in the gRPC client causing a divergence from current resource behavior.

Hi @ScottSuarez , In this resource we are using go rest client only right?
`clientAlloydb, err := alloydb.NewAlloyDBAdminRESTClient(c.context, option.WithHTTPClient(c.Client), option.WithUserAgent(userAgent), option.WithEndpoint(alloydbBasePath))`

Ref: `NewAlloyDBAdminRESTClient creates a new alloydb admin rest client.` from https://pkg.go.dev/cloud.google.com/go/alloydb/apiv1#NewAlloyDBAdminRESTClient

**Comment by ScottSuarez** [comment]
Hey @ravisiddhu , I spoke with my lead @rileykarson about this. We want to avoid adding the dependency on that client and have reverted the change which added it. Libraries in that repo are inconsistent given some are handwritten and grpc based. Could we change the implementation to use the methods @rileykarson outlined?

**Comment by ScottSuarez** [comment]
thanks ! sorry for any inconvenience !!

**Pull request closed without merging**